sycl/doc/Assert.md

```diff
@@ -0,0 +1,269 @@
+extensions.
+
+
+### Fallback approach
+
+If Device-side Runtime doesn't support `__devicelib_assert_fail` then a buffer
+based approach comes in place. The approach doesn't require any support from
+Device-side Runtime and Native Device Compiler. Neither it does from Low-level
+Runtime.
+
+Within this approach, a dedicated assert buffer is allocated and implicit kernel
+argument is introduced. The argument is an accessor with `discard_read_write`
+or `discard_write` access mode. Accessor metadata is stored to program scope
+variable. This allows to refer to the accessor without modifying each and every
+user's function. Fallback implementation of `__devicelib_assert_fail` restores
+accessor metadata from program scope variable and writes assert information to
+the assert buffer. Atomic operations are used in order to not overwrite existing
+information.
+
+DPCPP Runtime checks contents of the assert buffer for assert failure flag after
+kernel finishes.
+
+Both storing of accessor metadata and writing assert failure is performed with
+help of built-ins. Implementations of these builtins are substituted by
+frontend.
+
+#### Built-ins operation
+
+Accessor is a pointer augmented with offset and two ranges (access range and
+memory range).
+
+There are two built-ins provided by frontend:
+ * `__store_acc()` - to store accessor metadata into program-scope variable.
+ * `__store_assert_failure()` - to store flag about assert failure in a buffer
+   using the metadata stored in program-scope variable.
+
+The accessor should be stored to program scope variable in global address space
+using atomic operations. Motivation for using atomic operations: the program may
+contain several kernels and some of them could be running simultaneously on a
+single device.
+
+The `__store_assert_failure()` built-in atomically sets a flag in a buffer. The
+buffer is accessed using accessor metadata from program-scope variable. This
+built-in return a boolean value which is `true` if the flag is set by this call
+to `__store_assert_failure()` and `false` if the flag was already set.
+Motivation for using atomic operation is the same as with `__store_acc()`
+builtin.
+
+The following pseudo-code snippets shows how these built-ins are used.
+First of all, assume the following code as user's one:
+```
+void user_func(int X) {
+  assert(X && “X is nil”);
+}
+
+int main() {
+  queue Q(...);
+  Q.submit([&] (handler& CGH) {
+    CGH.single_task([=] () {
+      do_smth();
+      user_func(0);
+      do_smth_else();
+    });
+  });
+  ...
+}
+```
+
+The following LLVM IR pseudo code will be generated for the user's code:
+```
+@AssertBufferPtr = global void* null
+@AssertBufferAccessRange = ...
+@AssertBufferMemoryRange = ...
+@AssertBufferOffset = ...
+
+/// user's code
+void user_func(int X) {
+if (!(X && “X is nil")) {
+    __assert_fail(...);
+  }
+}
+
+users_kernel(...) {
+  do_smth()
+  user_func(0);
+  do_smth_else();
+}
+
+/// a wrapped user's kernel
```

Comment:
Could you please clarify at which level/in which component this wrapped kernel is being created? Will it be created somewhere in DPC++ headers or will it be generated by DPC++ FE?

Reply:
Added